wagtail/wagtail
# Rename `SkipLinkController` to `FocusController`, update unit tests & add story

## Description

The `SkipLinkController` was one of the first controllers implemented and as such has not adopted some of the practice that's emerged through other Stimulus adoption work.

## Details

This PR introduces the following:

1. A simpler name that reflects the **behaviour** not the **usage** of the Controller
2. Better unit tests with a structure that's aligned with many other unit testing
3. Better JSDoc leverage with proper `@example` tags
4. Abstract the common add/remove event listener behaviour and only remove the tabindex if the target element did not have one in the first place - pull this out to a robust util `forceFocus` as we use this in lots of other place in the code
5. Enhance the controller to be able to trigger focus on any targeted element, so we do not have to use it for links only, also ensure we can smooth scroll into view before triggering the focus for a better user experience
6. Add a Story that covers the example usage for additional manual testing
7. Add a Python unit test that covers the presence of the skip link, this is not every page but at least ensures we do not accidentally remove this code from the base admin template
8. Fix up an issue with our stubs/adapter files so we can better test scrolling with Jest function mocks

See also https://github.com/wagtail/wagtail/pull/12554 (a similar, in purpose, rename & clean up effort).

## How to test

Within the admin interface, once refreshed, press `tab`. The first target element should be the 'skip to main content' link, activating this (press enter) should move your focus away and into the main content. Pressing `tab` again, should focus on the first focusable element in the main content (e.g. the first link '34 pages' in the screen below).

<img width="1800" alt="Screenshot 2024-11-08 at 7 53 46 AM" src="https://github.com/user-attachments/assets/3dcf7944-cf92-498b-97e7-ff3919454556">

## Why this matters

Eventually we want to consider broader usage of the focus controller, either made available through documentation or for 'scroll to top' type buttons around longer page content. This is already a request for a different issue - https://github.com/wagtail/wagtail/issues/12128

This also makes it easier for us to eventually centralise the Jest testing for Stimulus controllers, the `await setup` pattern is super useful and makes adding future unit tests much easier.

Finally, the fact that the element's name (Link) is in the controller is a bit of an anti-pattern, any element could be used (e.g. a button) and should normally not be reflected in the controller name.

While renames seem unnecessary, without this rename it may seem like we need a different controller do do any 'focus' work, when in fact Stimulus shines when we create smaller modular 'behaviour' controllers and not usage specific ones. This aligns with our Stimulus tips, which were not really full formed when this controller was created. https://docs.wagtail.org/en/stable/contributing/ui_guidelines.html#helpful-tips
